core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java

```diff
@@ -949,6 +949,93 @@ public void modeInUnsupportedFormat() {
         exception.getMessage());
   }
 
+  private void testWeekOfYear(String date, int mode, int expectedResult) {
+    FunctionExpression expression = DSL
+        .week_of_year(DSL.literal(new ExprDateValue(date)), DSL.literal(mode));
+    assertEquals(INTEGER, expression.type());
+    assertEquals(String.format("week_of_year(DATE '%s', %d)", date, mode), expression.toString());
+    assertEquals(integerValue(expectedResult), eval(expression));
+  }
+
+  private void testNullMissingWeekOfYear(ExprCoreType date) {
+    when(nullRef.type()).thenReturn(date);
+    when(missingRef.type()).thenReturn(date);
+    assertEquals(nullValue(), eval(DSL.week_of_year(nullRef)));
+    assertEquals(missingValue(), eval(DSL.week_of_year(missingRef)));
+  }
+
+  @Test
+  public void invalid_week_of_year() {
+    testNullMissingWeekOfYear(DATE);
+    testNullMissingWeekOfYear(DATETIME);
+    testNullMissingWeekOfYear(TIMESTAMP);
+    testNullMissingWeekOfYear(STRING);
+
+    when(nullRef.type()).thenReturn(INTEGER);
+    when(missingRef.type()).thenReturn(INTEGER);
+    assertEquals(nullValue(), eval(DSL.week_of_year(DSL.literal("2019-01-05"), nullRef)));
+    assertEquals(missingValue(), eval(DSL.week_of_year(DSL.literal("2019-01-05"), missingRef)));
+
+    when(nullRef.type()).thenReturn(DATE);
+    when(missingRef.type()).thenReturn(INTEGER);
+    assertEquals(missingValue(), eval(DSL.week_of_year(nullRef, missingRef)));
+  }
+
+  @Test
+  public void week_of_year() {
+    testNullMissingWeekOfYear(DATE);
+    testNullMissingWeekOfYear(DATETIME);
+    testNullMissingWeekOfYear(TIMESTAMP);
+    testNullMissingWeekOfYear(STRING);
+
+    FunctionExpression expression = DSL
+        .week_of_year(DSL.literal(new ExprTimestampValue("2019-01-05 01:02:03")));
+    assertEquals(INTEGER, expression.type());
+    assertEquals("week_of_year(TIMESTAMP '2019-01-05 01:02:03')", expression.toString());
+    assertEquals(integerValue(0), eval(expression));
+
+    expression = DSL.week_of_year(DSL.literal("2019-01-05"));
+    assertEquals(INTEGER, expression.type());
+    assertEquals("week_of_year(\"2019-01-05\")", expression.toString());
+    assertEquals(integerValue(0), eval(expression));
+
+    expression = DSL.week_of_year(DSL.literal("2019-01-05 00:01:00"));
+    assertEquals(INTEGER, expression.type());
+    assertEquals("week_of_year(\"2019-01-05 00:01:00\")", expression.toString());
+    assertEquals(integerValue(0), eval(expression));
+
```

Comment:
I don't really know if we need all these tests, or even if they're correct.  Maybe a little of context why we sometimes get 52nd/53rd weeks of the month in January?

Reply:
I'll preface this by saying I did not write this test from scratch. It is based on the test for the `week` function above in the same file. I do think we need all the tests though, because they test the different behaviours of the `week` (and `week_of_year`) function based on the mode argument passed into the function.  Details can be found [here](https://www.w3resource.com/mysql/date-and-time-functions/mysql-week-function.php), but essentially the weird behaviour of getting a 52 or 53 has to do with how the days/weeks of the year are defined according to the mode. My understanding is that for these cases, the 1st week of the year is defined in a way that it doesn't contain January 1, so it falls back so that January 1 is in the last few weeks of the previous year.

I've tried to verify that the numbers in the the test are correct manually by just doing the calculation for the day of week according to the mode, and by running the same queries on this [tool](https://extendsclass.com/mysql-online.html). If you want to try and verify the numbers yourself so they make sense here's a picture of the weeks surrounding January 5, 2019 and the rules of the `week` function.
![image](https://user-images.githubusercontent.com/64448015/203830155-df7abd21-95be-4eb5-b7d5-678e4cf590b2.png)

Reply:
lol.  Thanks.  I think if we get the same results from SQL, then it's all good